bundles/org.openhab.binding.liquidcheck/src/main/java/org/openhab/binding/liquidcheck/internal/LiquidCheckHandler.java

```diff
@@ -0,0 +1,200 @@
+import org.openhab.core.thing.ThingStatusDetail;
+import org.openhab.core.thing.binding.BaseThingHandler;
+import org.openhab.core.types.Command;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The {@link LiquidCheckHandler} is responsible for handling commands, which are
+ * sent to one of the channels.
+ *
+ * @author Marcel Goerentz - Initial contribution
+ */
+@NonNullByDefault
+public class LiquidCheckHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(LiquidCheckHandler.class);
+    private final HttpClient httpClient;
+
+    private Map<String, String> oldProps = new HashMap<>();
+
+    private LiquidCheckConfiguration config = new LiquidCheckConfiguration();
+    private LiquidCheckHttpClient client = new LiquidCheckHttpClient(config, new HttpClient());
+
+    private @Nullable ScheduledFuture<?> polling;
+
+    public LiquidCheckHandler(Thing thing, HttpClient httpClient) {
+        super(thing);
+        this.httpClient = httpClient;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (channelUID.getId().equals(MEASURE_CHANNEL)) {
+            if (command instanceof OnOffType) {
+                try {
+                    if (client.isConnected()) {
+                        String response = client.measureCommand();
+                        CommData commandResponse = new Gson().fromJson(response, CommData.class);
+                        if (commandResponse != null) {
+                            if (!"success".equals(commandResponse.context.status)) {
+                                logger.warn("Starting the measurement was not successful!");
+                            }
+                        } else {
+                            logger.warn("The object commandResponse is null!");
+                        }
+                    }
+                } catch (TimeoutException | ExecutionException | JsonSyntaxException e) {
+                    logger.warn("This went wrong in handleCommand: {}", e.getMessage());
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+                updateState(channelUID, OnOffType.OFF);
+            }
+        }
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(LiquidCheckConfiguration.class);
+        oldProps = thing.getProperties();
+
+        updateStatus(ThingStatus.UNKNOWN);
+
+        // Example for background initialization:
+        scheduler.execute(() -> {
+            this.client = new LiquidCheckHttpClient(config, httpClient);
+            Initialize init = new Initialize(this.client);
+            while (!init.initialized) {
+                scheduler.scheduleWithFixedDelay(init, 0, config.refreshInterval, TimeUnit.SECONDS);
+            }
```

Comment:
I don't understand this loop. It seems you are quickly scheduling multiple new instances until first one is connected?

Reply:
No its trying to initialize in the refreshing rate. So if the user has set the refresh interval  to 1s then it will try to initialize every second.

Reply:
Okay, thanks, I understand the intention. However, the implementation is an infinite loop - you will schedule as many futures as you can, until one of them will initialize. If you add a logging statement in line 139:

```java
        @Override
        public void run() {
            logger.info("Initialize.run() called");
            boolean thingReachable = this.client.isConnected();
```

And configure the thing with hostname = "localhost" (or some other address which is not an expected device) you will see this:
```
2023-04-05 20:32:50.745 [INFO ] [uidcheck.internal.LiquidCheckHandler] - Initialize.run() called
2023-04-05 20:32:50.746 [INFO ] [uidcheck.internal.LiquidCheckHandler] - Initialize.run() called
2023-04-05 20:32:50.746 [INFO ] [uidcheck.internal.LiquidCheckHandler] - Initialize.run() called
2023-04-05 20:32:50.747 [INFO ] [uidcheck.internal.LiquidCheckHandler] - Initialize.run() called
2023-04-05 20:32:50.747 [INFO ] [uidcheck.internal.LiquidCheckHandler] - Initialize.run() called
2023-04-05 20:32:50.747 [INFO ] [uidcheck.internal.LiquidCheckHandler] - Initialize.run() called
2023-04-05 20:32:50.747 [INFO ] [uidcheck.internal.LiquidCheckHandler] - Initialize.run() called
```

You should schedule `init` only once, and also there is no need to do this inside `execute()`.

Perhaps it would be simpler to remove `Initialize` altogether and just schedule `PollingForData`?

Reply:
Okay, I got your point and will update the handler